### Student performance Indicator

#### Problem statement
This project understand how the student's performance (test score) is affected by other variables susch as Gender, Ethnicity, Parental level of education,lunch and test preparation course

#### Dataset information
dataset source - https://www.kaggle.com/datasets/spscientist/students-performance-in-exams.

dataset size - 8 columns and 1000 rows

###### Attributes -
###### gender - sex of student,
###### race_ethnicity - ethinisity of student(Group A,B,C,D,E)
###### parental_level_of_education	- Parents final education,
###### lunch - having lunch before test,
###### test_preparation_course - complete or not complete before test,
###### math_score ,
###### reading_score,
###### writing_score


In [1]:
## Import libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

In [4]:

df = pd.read_csv("C:/MLProjects/notebook/data/stud.csv")

In [6]:
## Top 5 rows
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [7]:
##shape of dataset
df.shape

(1000, 8)

#### Data checks to perform
###### Missing values, 
###### duplicates, datatype, 
###### number of unique values of each column, 
###### check statistics of data set, 
###### check various categories present in different categorical column 

In [13]:
print("Missing values : ", df.isna().sum())
print("There is no missing values present in the dataset.")

Missing values :  gender                         0
race_ethnicity                 0
parental_level_of_education    0
lunch                          0
test_preparation_course        0
math_score                     0
reading_score                  0
writing_score                  0
dtype: int64
There is no missing values present in the dataset.


In [14]:
print("Duplicated values: ", df.duplicated().sum())
print("There is no duplicate values present in the dataset.")


Duplicated values:  0
There is no duplicate values present in the dataset.


In [15]:
print("Datatype of columns: ", df.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race_ethnicity               1000 non-null   object
 2   parental_level_of_education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test_preparation_course      1000 non-null   object
 5   math_score                   1000 non-null   int64 
 6   reading_score                1000 non-null   int64 
 7   writing_score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB
Datatype of columns:  None


In [18]:
print("Unque columns: ", df.nunique())

Unque columns:  gender                          2
race_ethnicity                  5
parental_level_of_education     6
lunch                           2
test_preparation_course         2
math_score                     81
reading_score                  72
writing_score                  77
dtype: int64


In [19]:
print("statistics of dataset: ", df.describe())

statistics of dataset:         math_score  reading_score  writing_score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000


Insight - 
all means are very close to each other, i.e. between 66 to 68.05;

all standard deviation are closed to each other. i.e. between 14 to 15.19

min score for math is 0 in which for reading it is 17 and writing 10


### EDA

In [ ]:
## Categories in each column
print("Categories in 'gender' variable : ", end= " ")
print(df['gender'].unique())

print("Categories in 'race_ethnicity' variable: ", end = " ")
print(df['race_ethnicity'].unique())

print("Categories in 'parental_level_of_education' variable: ", end = " ")
print(df['parental_level_of_education'].unique())

print("Categories in 'lunch' variable: ", end = " ")
print(df['lunch'].unique())

print("Categories in 'test_preparation_course' variable: ", end = " ")
print(df['test_preparation_course'].unique())

print("Categories in 'math_score' variable: ", end = " ")
print(df['math_score'].unique())

print("Categories in 'reading_score' variable: ", end = " ")
print(df['reading_score'].unique())

print("Categories in 'writing_score' variable: ", end = " ")
print(df['writing_score'].unique())




Categories in 'gender' variable :  ['female' 'male']
Categories in 'race_ethnicity' variable:  ['group B' 'group C' 'group A' 'group D' 'group E']
Categories in 'parental_level_of_education' variable:  ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
Categories in 'lunch' variable:  ['standard' 'free/reduced']


In [20]:
df.columns

Index(['gender', 'race_ethnicity', 'parental_level_of_education', 'lunch',
       'test_preparation_course', 'math_score', 'reading_score',
       'writing_score'],
      dtype='object')

In [32]:
### Defining numerical and categorical features
num_features = [feature for feature in df.columns if df[feature].dtype != 'O']
cat_features = [feature for feature in df.columns if df[feature].dtype == 'O']

print("We have {} numerical features : {}".format(len(num_features),(num_features)))
print("\nWe have {} Categorical features : {}".format(len(cat_features),(cat_features)))

We have 3 numerical features : ['math_score', 'reading_score', 'writing_score']

We have 5 Categorical features : ['gender', 'race_ethnicity', 'parental_level_of_education', 'lunch', 'test_preparation_course']


### Adding columns for total score and average score

In [36]:
df['total score'] = df['math_score'] + df['reading_score'] + df['writing_score']
df['average score'] = df['total score'] / 3
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   total score  average score  
0          218      72.666667  
1          247      82.333333  
2          278      92.666667  
3          148      49.333333  
4          229      76.333333

In [45]:
## students with full marks in 

read_full = df[df['reading_score'] == 100]['average score'].count()
write_full = df[df['writing_score'] == 100]['average score'].count()
math_full = df[df['math_score'] == 100]['average score'].count()

print(f"Number of students with full marks in reading: {read_full}",)
print(f"Number of students with full marks in writing: {write_full}",)
print(f"Number of students with full marks in math: {math_full}",)


Number of students with full marks in reading: 17
Number of students with full marks in writing: 14
Number of students with full marks in math: 7


In [48]:
read_less_20 = df[df['reading_score'] <= 20]['average score'].count()
write_less_20 = df[df['writing_score'] <= 20]['average score'].count()
math_less_20 = df[df['math_score'] <= 20]['average score'].count()

print(f"Number of students less than 20 marks in reading: {read_less_20}",)
print(f"Number of students less than 20 marks in writing: {write_less_20}",)
print(f"Number of students less than 20 marks in math: {math_less_20}",)

Number of students less than 20 marks in reading: 1
Number of students less than 20 marks in writing: 3
Number of students less than 20 marks in math: 4


insight - From above analysis we can conclude that student performed worst in math but best in reading section.